src/Controller/OPDS/OpdsController.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+namespace App\Controller\OPDS;
+
+use ACSEO\TypesenseBundle\Finder\SpecificCollectionFinder;
+use App\Entity\OpdsAccess;
+use App\Entity\Shelf;
+use App\Entity\User;
+use App\OPDS\Opds;
+use App\Repository\BookInteractionRepository;
+use App\Repository\BookRepository;
+use App\Repository\ShelfRepository;
+use App\Service\ShelfManager;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+#[Route('/opds/{accessKey}', name: 'opds_')]
+class OpdsController extends AbstractController
+{
+    public function __construct(private readonly Opds $opds, private readonly BookRepository $bookRepository, private readonly SpecificCollectionFinder $bookFinder)
+    {
+    }
+
+    private function generateOpdsUrl(string $route, array $params = []): string
+    {
+        $user = $this->getUser();
+        if (!$user instanceof User) {
+            throw $this->createAccessDeniedException('Invalid user');
+        }
+
+        $accesses = $user->getOpdsAccesses()->toArray();
+        if (count($accesses) === 0) {
+            throw $this->createAccessDeniedException('Invalid user, no opds access');
+        }
+        /** @var OpdsAccess $opdsAccess */
+        $opdsAccess = reset($accesses);
+
+        $params['accessKey'] = $params['accessKey'] ?? $opdsAccess->getToken();
+
+        return $this->generateUrl($route, $params, UrlGeneratorInterface::ABSOLUTE_URL);
+    }
+
+    #[Route('/', name: 'start')]
+    public function index(Request $request): Response
+    {
+        $opds = $this->opds->getOpdsConfig(''.$request->get('accessKey'));
```

Comment:
You have 5 occurrences for this line.

1. Create a private function
2. Do not do hack to get a string but use  `$request->something->getString('accessKey')` , something is either the attribute, the request or the query..
3. I would rename `getOpdsConfig` to `getOpdsConfigFromAccessKey`